Fix comparison of machine addresses with openstack instance addresses

mdbooth opened the pull request:
Firstly this test was broken because skipUnlessMachineAPIOperator is broken and was executing against a UPI cluster. We address this by only fetching OpenStack instances for machines which exist. If no machines exist (because it's UPI) the test trivially succeeds.

Secondly we improve the error reporting. We update the gomega matchers to operate on the actual data. This means that if they don't match the non-matching data will be included in the error message.

pierreprinetti commented on test/extended/openstack/servers.go:
```diff
@@ -302,3 +315,10 @@ func skipUnlessMachineAPIOperator(dc dynamic.Interface, c coreclient.NamespaceIn
 	o.Expect(err).NotTo(o.HaveOccurred())
 }
+
+func mapKeys[K comparable, V any](m map[K]V) (keys []K) {
```
first generics in openstack-test!

pierreprinetti commented:
Your solution for this test looks fine, however it seems a bit too implicit WRT the UPI case in particular.
The crux of it all being:
> Incidentally `skipUnlessMachineAPIOperator` seems broken

Do you see a way to address this more broadly? Maybe by making `skipUnlessMachineAPIOperator` skip if there are no ControlPlane machines in the cluster? Or maybe, UPI should have a manifest deactivating MAO in CVO and we check that here?

mdbooth commented:
> Your solution for this test looks fine, however it seems a bit too implicit WRT the UPI case in particular. The crux of it all being:
> 
> > Incidentally `skipUnlessMachineAPIOperator` seems broken
> 
> Do you see a way to address this more broadly? Maybe by making `skipUnlessMachineAPIOperator` skip if there are no ControlPlane machines in the cluster? Or maybe, UPI should have a manifest deactivating MAO in CVO and we check that here?

Honestly I've always preferred code which handles empty lists by trivially iterating over them rather than adding a special case. This test asserts that all openshift machines must have corresponding openstack instances with the same addresses. This is true for both UPI and IPI. That's extremely tidy to me.

I haven't looked at other uses of skipUnlessMachineAPIOperator yet. We will need to fix it properly if we have any tests that genuinely need to special case it, but perhaps we can even delete it? Less special cases is more better. Lets treat them one by one.

pierreprinetti commented:
> is test asserts that all openshift machines must have corresponding openstack instances with the same addresses.

That's what the test does, not what it says ("The OpenStack platform - on instance creation - should include the addresses on the machine specs")

mdbooth commented:
> > is test asserts that all openshift machines must have corresponding openstack instances with the same addresses.
> 
> That's what the test does, not what it says ("The OpenStack platform - on instance creation - should include the addresses on the machine specs")

Good point. I fixed the test description to describe what it's actually doing. This hasn't changed, btw.

I also added a comment pointing out that it works for UPI too.

pierreprinetti commented:
Thank you.
/lgtm

mdbooth commented:
Verify is failing, but I don't know what the output means:

```
> ./hack/verify.sh


Jan 17 18:50:19.872: INFO: Enabling in-tree volume drivers
758c758
<       "[sig-installer][Suite:openshift/openstack] The OpenStack platform on instance creation should include the addresses on the machine specs": "",
---
>       "[sig-installer][Suite:openshift/openstack] The OpenStack platform on instance creation should report all openstack instance addresses on the corresponding Machine object": "",
```

Is it complaining that the test name changed?

mdbooth commented:
> Verify is failing, but I don't know what the output means:
> 
> ```
> > ./hack/verify.sh
> 
> 
> Jan 17 18:50:19.872: INFO: Enabling in-tree volume drivers
> 758c758
> <       "[sig-installer][Suite:openshift/openstack] The OpenStack platform on instance creation should include the addresses on the machine specs": "",
> ---
> >       "[sig-installer][Suite:openshift/openstack] The OpenStack platform on instance creation should report all openstack instance addresses on the corresponding Machine object": "",
> ```
> 
> Is it complaining that the test name changed?

Found it. It's complaining that the generated file is out of date. I'm going to add a better error message along the lines of the gofmt one.

pierreprinetti commented on hack/verify.sh:
```diff
@@ -27,7 +27,14 @@ verify_generated() {
-	diff "$old" "$new"
+	if ! diff "$old" "$new"; then
+            echo 'Generated files are out of date'
+            echo 'Run "Go generate ./..."'
```
```
echo 'Run "make update"'

mdbooth commented:
/cherry-pick release-4.12
